src/text/non_blocking/raw_text_reader.rs

```diff
@@ -723,47 +878,36 @@ impl<A: AsRef<[u8]>> IonReader for RawTextReader<A> {
     }
 
     fn step_out(&mut self) -> IonResult<()> {
-        if self.parents.is_empty() {
-            return illegal_operation(
-                "Cannot call `step_out()` when the reader is at the top level.",
-            );
-        }
-
-        // The container we're stepping out of.
-        let parent = self.parents.last().unwrap();
-
-        // If we're not at the end of the current container, advance the cursor until we are.
-        // Unlike the binary reader, which can skip-scan, the text reader must visit every value
-        // between its current position and the end of the container.
-        if !parent.is_exhausted() {
-            while let RawStreamItem::Value(_) | RawStreamItem::Null(_) = self.next()? {}
-        }
-
-        // Remove the parent container from the stack and clear the current value.
-        let _ = self.parents.pop();
-        self.current_value = None;
-
-        if self.parents.is_empty() {
-            // We're at the top level; nothing left to do.
-            return Ok(());
-        }
-
-        // We've stepped out, but the reader isn't at the top level. We're still inside another
-        // container. Make sure the container was followed by either the appropriate delimiter
-        // or the end of its parent.
-        let container_type = self.parents.last().unwrap().ion_type();
-        match container_type {
-            IonType::List => {
-                self.parse_expected("list delimiter or end", list_delimiter)?;
-            }
-            IonType::SExpression => {
-                self.parse_expected("s-expression delimiter or end", s_expression_delimiter)?;
-            }
-            IonType::Struct => {
-                self.parse_expected("struct delimiter or end", struct_delimiter)?;
+        // If an incomplete error occurs during a step_out, we will re-enter step_out and:
+        //    If we have any error other than ContainerDelimiter, we need to finish exhausting
+        //    the current container and complete the step_out that failed.
+        //
+        //    Once the failed step_out is done, we need to "bubble up" and step_out of any of
```

Comment:
I don't think I understand this bit -- let's talk through it offline.

Reply:
👍